[llvm] Ensure that soft float targets don't use float/vector code for memops.

alexrp opened the pull request:
The RISC-V backend still gets this very wrong, hence why I'm not adding a test for it. In its case, the reason appears to be that it doesn't support these attributes at all (yet?). This will require much deeper surgery to fix, which I might do in follow-up patches.

Closes #105978.

alexrp commented on llvm/lib/Target/Mips/MipsISelLowering.cpp:
```diff
@@ -4331,14 +4331,6 @@ MipsTargetLowering::isOffsetFoldingLegal(const GlobalAddressSDNode *GA) const {
   return false;
 }
 
-EVT MipsTargetLowering::getOptimalMemOpType(
```
This hunk is a leftover from a previous version of the patch; please ignore when reviewing. I'll restore this.

(The implementation looks a bit suspect in how it ignores alignment completely and doesn't check the `strict-align` target feature. But that's a completely separate issue.)

arichardson replied:
MIPS can use pairs of LWL/LWR LDL/LDR for unaligned accesses, so i64/i32 is always legal here even if unaligned accesses are not possible. (As long as the lowering takes this into account, which I believe it does).

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -8846,9 +8846,14 @@ static bool findGISelOptimalMemOpLowering(std::vector<LLT> &MemOps,
-  LLT Ty = TLI.getOptimalMemOpLLT(Op, FuncAttributes);
-
-  if (Ty == LLT()) {
+  bool WantIntScalar = TLI.useSoftFloat() ||
+                       FuncAttributes.hasFnAttr(Attribute::NoImplicitFloat);
```
I'd prefer to avoid eagerly querying these edge cases. Also this seems like another case where we have overlapping and poorly defined mechanisms. For example useSoftFloat is a virtual function without any arguments, which is kind of useless. Why isn't it a simple function field, and isn't it a function of the use-soft-float attribute, so it can't just be part of the subtarget's TargetLowering? 

At least could just merge these into a helper for implementations of getOptimalMemOpLLT to use?

alexrp replied:
@arsenm ping for clarification on this point:

> Do you mean that you would prefer if I just fixed each `getOptimalMemOpLLT()`/`getOptimalMemOpType()` implementation to check for these attributes when they want to potentially return a float/vector type?

arsenm replied:
Yes. Can put the query into a helper somewhere for the implementations to use?

alexrp committed:
[llvm] Ensure that soft float targets don't use float/vector code for memops.

Note: The RISC-V backend still gets this very wrong, hence why I'm not adding a
test for it. In its case, the reason appears to be that it doesn't support these
attributes at all (yet?). This will require much deeper surgery to fix, which I
might do in follow-up patches.

Closes #105978.

arsenm commented on llvm/test/CodeGen/ARM/memop-soft-float.ll:
```diff
@@ -0,0 +1,128 @@
+; ARM-NEON-NEXT:    str r12, [r12]
+; ARM-NEON-NEXT:    pop {r4, lr}
+; ARM-NEON-NEXT:    mov pc, lr
+    call void @llvm.memcpy.p0.p0.i32(ptr null, ptr null, i32 32, i1 true)
+    call void @llvm.memset.p0.i32(ptr null, i8 0, i32 32, i1 true)
+    ret void
```
2 space indent. Also avoid the UB null accesses?

arsenm commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -2023,6 +2023,11 @@ class TargetLoweringBase {
     return LLT();
   }
 
+  bool useIntScalarMemOps(const AttributeList &FuncAttributes) const {
```
Needs documentation. Also the name should reflect the meaning, not the implementation. This seems to me like it's just an extension of useSoftFloat which contains the function context. Can you just add the argument to useSoftFloat instead?

alexrp replied:
> Also the name should reflect the meaning, not the implementation.

Hmm, I'm not sure I follow; it seems to me that's what it does already? What do you have in mind?

> This seems to me like it's just an extension of useSoftFloat which contains the function context. Can you just add the argument to useSoftFloat instead?

The backends are inconsistent about whether they use `TM.Options.FloatABIType` or `useSoftFloat()` for ABI questions. The ideal (IMO) would be that they all agree on using the former for ABI questions and the latter for "can we use float/vector code within ABI boundaries?" (which would just be a question of the `noimplicitfloat` attribute, at least for now).

This PR is intended to be an incremental step towards that ideal. But if you think it makes more sense, I can try to tackle that whole inconsistency in one go if you think that makes more sense.

arsenm replied:
"useIntScalarMemOps" Why int? Why scalar? What mem ops? In what context? The core change has nothing to do with any of these things, this is about handling of some fuzzy definition of floating point operations which I've never understood

arsenm replied:
useSoftFloat is better but also vague. use "soft" float for what?

alexrp replied:
> "useIntScalarMemOps" Why int? Why scalar? What mem ops? In what context? The core change has nothing to do with any of these things

The `IntScalar` term is used in other places in the repo and just seems to mean non-vector integers. Granted, the `Mem` part is unnecessary because the function could equally well apply to other areas than just `getOptimalMemOpType()` which I'm modifying in this PR.

Essentially the question being asked is: Are we able/allowed to use float/vector instructions? If `useSoftFloat()` is true (usually a result of CPU feature flags), then the answer is *technically* yes but it would be pointless work because the instructions would just be converted to soft float equivalents immediately after, and likely result in worse code than had we just used the int-scalar path to begin with. If `noimplicitfloat` is set on the function, then the answer is a definite no, probably because we're compiling kernel code or similar.

So, perhaps `avoidFloatOrVectorOps()` or something like that?

> useSoftFloat is better but also vague. use "soft" float for what?

Well, indeed, that's the problem I was alluding to above re: how backends are inconsistent about what they take it to mean.

---

I also forgot to mention the `use-soft-float` function attribute, which each backend typically transforms into a subtarget CPU feature flag (like `soft-float` or `hard-float`) which finally informs `useSoftFloat()`. It's all quite confusing and messy.

uweigand replied:
As I see it, the term "soft float" has traditionally been used for one of two distinct concepts:
1) Hardware features associated with floating-point (and/or vector) are not available.  This includes instructions, but more importantly also *registers*.   These may not be available because the HW doesn't actually have them, or because they must not be used in the given context (e.g. Linux kernel code).
2) The ABI for passing floating-point (and/or vector) types does not use any floating-point (vector) HW features (specifically, registers).

Now clearly if we have no float/vector registers, we cannot use an ABI that requires them.   The inverse is not necessarily true - you could implement an ABI that does not use float registers, but the generated code is still allowed to use float registers within a function.  That's the case that I think sometimes causes confusion.

At least on SystemZ, the intent of the `-msoft-float` flag in all its permutations has always been both 1) and 2) - i.e. it implements a non-standard ABI, and also prohibits any use of floating-point or vector registers.

alexrp replied:
> The inverse is not necessarily true - you could implement an ABI that does not use float registers, but the generated code is still allowed to use float registers within a function. That's the case that I think sometimes causes confusion.

I think ARM is the only backend that supports this configuration right now, although it is pretty fragile if you don't massage the CPU feature flags and function attributes just right IIRC. I think ideally every backend should allow this configuration, though; not doing so seems mostly like an artificial restriction.

In my ideal world, you'd set `TM.Options.FloatABIType` to pick the soft/hard ABI, and in the case of soft ABI, optionally set the `use-soft-float` function attribute if you want to also lower all float/vector computation to soft float. `noimplicitfloat` would keep keep its current meaning, i.e. the compiler isn't allowed to introduce float/vector operations that weren't already there.

Clang options would then translate like this:

* `-mfloat-abi=soft`/`-msoft-float`: `FloatABI::ABIType::Soft` + `use-soft-float` + `noimplicitfloat`
* `-mfloat-abi=softfp`: `FloatABI::ABIType::Soft`
* `-mfloat-abi=hard`/`-mhard-float`: `FloatABI::ABIType::Hard`
* `-mno-implicit-float`: `noimplicitfloat`

uweigand replied:
Just one additional comment: "if you want to also lower all float/vector computation to soft float" is not the actual concern for the primary use case of `-msoft-float` on our platform: the Linux kernel.   This code base already never uses any float/vector computation in the first place.  The primary concern is to prevent the compiler from introducing uses of float/vector registers even for code that does have any float/vector operations at the source level.

uweigand commented on llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:
```diff
@@ -1440,9 +1440,11 @@ bool SystemZTargetLowering::findOptimalMemOpLowering(
-  return Subtarget.hasVector() ? MVT::v2i64 : MVT::Other;
+EVT SystemZTargetLowering::getOptimalMemOpType(
+    const MemOp &Op, const AttributeList &FuncAttributes) const {
+  return Subtarget.hasVector() && !useIntScalarMemOps(FuncAttributes)
```
Not sure why this is necessary.  This is supposed to handled once and for all in `SystemZSubtarget::initializeSubtargetDependencies` by switching off the vector facility completely for soft-float:
```
  // -msoft-float implies -mno-vx.
  if (HasSoftFloat)
    HasVector = false;
```

alexrp replied:
I think you might be right that it's not needed. I don't actually recall why I modified this function in the initial version of the PR.

I'll re-check and remove it if it's unnecessary.
